File: backend/app/data_sources/fundamentals.py
"""Greenblatt-style fundamental screening via yfinance data."""

import logging
from app.data_sources.yahoo_finance import fetch_info_safe

logger = logging.getLogger(__name__)


def fetch_fundamentals(tickers: list[str]) -> dict:
    """Fetch fundamental metrics for a list of tickers.

    yfinance .info already provides all needed fields:
    freeCashflow, netIncomeToCommon, totalDebt, totalCash, enterpriseValue, etc.
    """
    results = {}
    for ticker_sym in tickers:
        try:
            info = fetch_info_safe(ticker_sym)
            results[ticker_sym] = _extract_fundamentals(info, ticker_sym)
        except Exception as e:
            logger.warning(f"Fundamentals failed for {ticker_sym}: {e}")
            results[ticker_sym] = {"error": str(e)}
    return results


def _extract_fundamentals(info: dict, ticker: str) -> dict:
    """Extract all fundamental metrics from yfinance info dict."""
    valuation = _compute_valuation(info)
    quality = _compute_quality(info)
    growth = _compute_growth(info)
    health = _compute_health(info)
    flag, reasoning = _generate_valuation_flag(valuation, quality, growth, health)

    return {
        "ticker": ticker,
        "valuation": valuation,
        "quality": quality,
        "growth": growth,
        "health": health,
        "valuation_flag": flag,
        "flag_reasoning": reasoning,
    }


def _compute_valuation(info: dict) -> dict:
    """P/E, EV/EBIT, earnings yield, FCF yield."""
    pe = info.get("trailingPE")
    forward_pe = info.get("forwardPE")
    ev = info.get("enterpriseValue")
    ebitda = info.get("ebitda")
    market_cap = info.get("marketCap")
    fcf = info.get("freeCashflow")

    ev_ebit = round(ev / ebitda, 2) if ev and ebitda and ebitda > 0 else None
    earnings_yield = round(1 / pe * 100, 2) if pe and pe > 0 else None
    fcf_yield = round(fcf / market_cap * 100, 2) if fcf and market_cap and market_cap > 0 else None
    price_to_book = info.get("priceToBook")

    return {
        "pe_ratio": round(pe, 2) if pe else None,
        "forward_pe": round(forward_pe, 2) if forward_pe else None,
        "ev_ebit": ev_ebit,
        "earnings_yield": earnings_yield,
        "fcf_yield": fcf_yield,
        "price_to_book": round(price_to_book, 2) if price_to_book else None,
    }


def _compute_quality(info: dict) -> dict:
    """ROIC, ROE, profit margin."""
    roe = info.get("returnOnEquity")
    profit_margin = info.get("profitMargins")

    # Approximate ROIC: net income / (total equity + total debt - cash)
    net_income = info.get("netIncomeToCommon")
    total_debt = info.get("totalDebt", 0)
    total_cash = info.get("totalCash", 0)
    book_value = info.get("bookValue")
    shares = info.get("sharesOutstanding")

    roic = None
    if net_income and book_value and shares:
        total_equity = book_value * shares
        invested_capital = total_equity + (total_debt or 0) - (total_cash or 0)
        if invested_capital > 0:
            roic = round(net_income / invested_capital * 100, 2)

    return {
        "roic": roic,
        "roe": round(roe * 100, 2) if roe else None,
        "profit_margin": round(profit_margin * 100, 2) if profit_margin else None,
    }


def _compute_growth(info: dict) -> dict:
    """Revenue and earnings growth."""
    return {
        "revenue_growth": _pct(info.get("revenueGrowth")),
        "earnings_growth": _pct(info.get("earningsGrowth")),
        "earnings_quarterly_growth": _pct(info.get("earningsQuarterlyGrowth")),
    }


def _compute_health(info: dict) -> dict:
    """Debt/Equity, current ratio."""
    de = info.get("debtToEquity")
    return {
        "debt_to_equity": round(de / 100, 2) if de else None,  # yfinance returns as percentage
        "current_ratio": _safe_round(info.get("currentRatio")),
        "quick_ratio": _safe_round(info.get("quickRatio")),
    }


def _generate_valuation_flag(
    valuation: dict, quality: dict, growth: dict, health: dict
) -> tuple[str, str]:
    """Greenblatt-style scoring for valuation flag."""
    score = 0
    reasons = []

    # Low EV/EBIT is attractive
    ev_ebit = valuation.get("ev_ebit")
    if ev_ebit is not None:
        if ev_ebit < 10:
            score += 2
            reasons.append(f"Low EV/EBIT ({ev_ebit}x)")
        elif ev_ebit < 15:
            score += 1
        elif ev_ebit > 25:
            score -= 1
            reasons.append(f"High EV/EBIT ({ev_ebit}x)")
        elif ev_ebit > 35:
            score -= 2
            reasons.append(f"Very high EV/EBIT ({ev_ebit}x)")

    # High ROIC is quality
    roic = quality.get("roic")
    if roic is not None:
        if roic > 20:
            score += 2
            reasons.append(f"Strong ROIC ({roic}%)")
        elif roic > 12:
            score += 1
        elif roic < 5:
            score -= 1
            reasons.append(f"Weak ROIC ({roic}%)")

    # FCF yield
    fcf_yield = valuation.get("fcf_yield")
    if fcf_yield is not None:
        if fcf_yield > 8:
            score += 1
            reasons.append(f"High FCF yield ({fcf_yield}%)")
        elif fcf_yield < 0:
            score -= 1
            reasons.append(f"Negative FCF yield ({fcf_yield}%)")

    # Earnings growth
    eg = growth.get("earnings_growth")
    if eg is not None:
        if eg > 15:
            score += 1
        elif eg < -10:
            score -= 1

    # Debt check
    de = health.get("debt_to_equity")
    if de is not None and de > 2.0:
        score -= 1
        reasons.append(f"High leverage (D/E {de})")

    if score >= 3:
        flag = "UNDERVALUED_OPPORTUNITY"
    elif score <= -2:
        flag = "OVERVALUATION_WARNING"
    else:
        flag = "FAIRLY_VALUED"

    reasoning = "; ".join(reasons) if reasons else "Mixed signals across metrics"
    return flag, reasoning


def _pct(val) -> float | None:
    return round(val * 100, 2) if val is not None else None


def _safe_round(val, digits=2) -> float | None:
    return round(val, digits) if val is not None else None

File: laravel-backend/app/DataSources/Fundamentals.php
<?php

namespace App\DataSources;

use Illuminate\Support\Facades\Log;

/**
 * Greenblatt-style fundamental screening via Massive.com data.
 */
class Fundamentals
{
    private YahooFinance $yahooFinance;

    public function __construct(YahooFinance $yahooFinance)
    {
        $this->yahooFinance = $yahooFinance;
    }

    public function fetchFundamentals(array $tickers): array
    {
        $results = [];
        foreach ($tickers as $ticker) {
            try {
                $info = $this->yahooFinance->fetchInfoSafe($ticker);
                $results[$ticker] = $this->extractFundamentals($info, $ticker);
            } catch (\Throwable $e) {
                Log::warning("Fundamentals failed for {$ticker}: {$e->getMessage()}");
                $results[$ticker] = ['error' => $e->getMessage()];
            }
        }
        return $results;
    }

    private function extractFundamentals(array $info, string $ticker): array
    {
        $valuation = $this->computeValuation($info);
        $quality = $this->computeQuality($info);
        $growth = $this->computeGrowth($info);
        $health = $this->computeHealth($info);
        [$flag, $reasoning] = $this->generateValuationFlag($valuation, $quality, $growth, $health);

        return [
            'ticker' => $ticker,
            'valuation' => $valuation,
            'quality' => $quality,
            'growth' => $growth,
            'health' => $health,
            'valuation_flag' => $flag,
            'flag_reasoning' => $reasoning,
        ];
    }

    private function computeValuation(array $info): array
    {
        $pe = $info['trailingPE'] ?? null;
        $forwardPe = $info['forwardPE'] ?? null;
        $ev = $info['enterpriseValue'] ?? null;
        $ebitda = $info['ebitda'] ?? null;
        $marketCap = $info['marketCap'] ?? null;
        $fcf = $info['freeCashflow'] ?? null;
        $priceToBook = $info['priceToBook'] ?? null;

        $evEbit = ($ev && $ebitda && $ebitda > 0) ? round($ev / $ebitda, 2) : null;
        $earningsYield = ($pe && $pe > 0) ? round(1 / $pe * 100, 2) : null;
        $fcfYield = ($fcf && $marketCap && $marketCap > 0) ? round($fcf / $marketCap * 100, 2) : null;

        return [
            'pe_ratio' => $pe ? round($pe, 2) : null,
            'forward_pe' => $forwardPe ? round($forwardPe, 2) : null,
            'ev_ebit' => $evEbit,
            'earnings_yield' => $earningsYield,
            'fcf_yield' => $fcfYield,
            'price_to_book' => $priceToBook ? round($priceToBook, 2) : null,
        ];
    }

    private function computeQuality(array $info): array
    {
        $roe = $info['returnOnEquity'] ?? null;
        $profitMargin = $info['profitMargins'] ?? null;

        $netIncome = $info['netIncomeToCommon'] ?? null;
        $totalDebt = $info['totalDebt'] ?? 0;
        $totalCash = $info['totalCash'] ?? 0;
        $bookValue = $info['bookValue'] ?? null;
        $shares = $info['sharesOutstanding'] ?? null;

        $roic = null;
        if ($netIncome && $bookValue && $shares) {
            $totalEquity = $bookValue * $shares;
            $investedCapital = $totalEquity + ($totalDebt ?? 0) - ($totalCash ?? 0);
            if ($investedCapital > 0) {
                $roic = round($netIncome / $investedCapital * 100, 2);
            }
        }

        return [
            'roic' => $roic,
            'roe' => $roe ? round($roe * 100, 2) : null,
            'profit_margin' => $profitMargin ? round($profitMargin * 100, 2) : null,
        ];
    }

    private function computeGrowth(array $info): array
    {
        return [
            'revenue_growth' => $this->pct($info['revenueGrowth'] ?? null),
            'earnings_growth' => $this->pct($info['earningsGrowth'] ?? null),
            'earnings_quarterly_growth' => $this->pct($info['earningsQuarterlyGrowth'] ?? null),
        ];
    }

    private function computeHealth(array $info): array
    {
        $de = $info['debtToEquity'] ?? null;
        return [
            'debt_to_equity' => $de ? round($de / 100, 2) : null,
            'current_ratio' => isset($info['currentRatio']) ? round($info['currentRatio'], 2) : null,
            'quick_ratio' => isset($info['quickRatio']) ? round($info['quickRatio'], 2) : null,
        ];
    }

    private function generateValuationFlag(array $valuation, array $quality, array $growth, array $health): array
    {
        $score = 0;
        $reasons = [];

        $evEbit = $valuation['ev_ebit'] ?? null;
        if ($evEbit !== null) {
            if ($evEbit < 10) {
                $score += 2;
                $reasons[] = "Low EV/EBIT ({$evEbit}x)";
            } elseif ($evEbit < 15) {
                $score += 1;
            } elseif ($evEbit > 35) {
                $score -= 2;
                $reasons[] = "Very high EV/EBIT ({$evEbit}x)";
            } elseif ($evEbit > 25) {
                $score -= 1;
                $reasons[] = "High EV/EBIT ({$evEbit}x)";
            }
        }

        $roic = $quality['roic'] ?? null;
        if ($roic !== null) {
            if ($roic > 20) {
                $score += 2;
                $reasons[] = "Strong ROIC ({$roic}%)";
            } elseif ($roic > 12) {
                $score += 1;
            } elseif ($roic < 5) {
                $score -= 1;
                $reasons[] = "Weak ROIC ({$roic}%)";
            }
        }

        $fcfYield = $valuation['fcf_yield'] ?? null;
        if ($fcfYield !== null) {
            if ($fcfYield > 8) {
                $score += 1;
                $reasons[] = "High FCF yield ({$fcfYield}%)";
            } elseif ($fcfYield < 0) {
                $score -= 1;
                $reasons[] = "Negative FCF yield ({$fcfYield}%)";
            }
        }

        $eg = $growth['earnings_growth'] ?? null;
        if ($eg !== null) {
            if ($eg > 15) $score += 1;
            elseif ($eg < -10) $score -= 1;
        }

        $de = $health['debt_to_equity'] ?? null;
        if ($de !== null && $de > 2.0) {
            $score -= 1;
            $reasons[] = "High leverage (D/E {$de})";
        }

        if ($score >= 3) {
            $flag = 'UNDERVALUED_OPPORTUNITY';
        } elseif ($score <= -2) {
            $flag = 'OVERVALUATION_WARNING';
        } else {
            $flag = 'FAIRLY_VALUED';
        }

        $reasoning = ! empty($reasons) ? implode('; ', $reasons) : 'Mixed signals across metrics';
        return [$flag, $reasoning];
    }

    private function pct(?float $val): ?float
    {
        return $val !== null ? round($val * 100, 2) : null;
    }
}
